File: 0018.4Sum/solution.py
from typing import List


class Solution:
    def four_sum(self, nums: List[int], target: int = 0) -> List[List[int]]:

        nums = sorted(nums)
        res = []

        for i in range(len(nums) - 3):

            if i > 0 and nums[i] == nums[i - 1]:  # 移除首位相同的项
                continue

            if sum(nums[:3], nums[i]) > target:
                break

            if sum(nums[-3:], nums[i]) < target:
                continue

            for j in range(i + 1, len(nums) - 2):

                if j - i > 1 and nums[j] == nums[j - 1]:  # 移除第二位相同的项
                    continue

                if sum(nums[j : j + 3], nums[i]) > target:
                    break

                if sum([nums[i], nums[j], nums[-1], nums[-2]]) < target:
                    continue

                left, right = j + 1, len(nums) - 1

                while left < right:
                    s = nums[i] + nums[j] + nums[left] + nums[right]

                    if s < target:
                        left += 1
                    elif s > target:
                        right -= 1
                    else:
                        res.append([nums[i], nums[j], nums[left], nums[right]])
                        left += 1
                        right -= 1

                        while left < right and nums[left] == nums[left - 1]:
                            left += 1
                        while left < right and nums[right] == nums[right + 1]:
                            right -= 1
        return res


if __name__ == "__main__":
    assert Solution().four_sum([0, 0, 0, 0]) == [[0, 0, 0, 0]]

File: 0018.4Sum/solution.go
package main

import (
	"fmt"
	"sort"
)

func fourSum(nums []int, target int) (ans [][]int) {
	if len(nums) < 4 {
		return ans
	}
	sort.Ints(nums)
	for m := 0; m < len(nums)-3; m++ {
		if m > 0 && nums[m-1] == nums[m] {
			continue
		}
		for i := m + 1; i < len(nums)-2; i++ {
			if i-m > 1 && nums[i] == nums[i-1] { // distinct
				continue
			}
			l, r := i+1, len(nums)-1
			for l < r {
				if l > i+1 && nums[l] == nums[l-1] { // distinct
					l++
					continue
				}
				if r < len(nums)-2 && nums[r] == nums[r+1] { // distinct
					r--
					continue
				}
				sum := nums[m] + nums[i] + nums[l] + nums[r]

				if sum < target {
					l++
				} else if sum > target {
					r--
				} else {
					ans = append(ans, []int{nums[m], nums[i], nums[l], nums[r]})
					l++
					r--
				}
			}
		}
	}
	return ans
}

func main() {
	fmt.Println(fourSum([]int{1, 0, -1, 0, -2, 2}, 0))
}
